In [125]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.cluster import KMeans

In [126]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [127]:
# Generate summary statistics
# df_market_data.columns
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [128]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [129]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
df_market_data_scaled = StandardScaler().fit_transform(df_market_data[['price_change_percentage_24h', 'price_change_percentage_7d',\
       'price_change_percentage_14d', 'price_change_percentage_30d',\
       'price_change_percentage_60d', 'price_change_percentage_200d',\
       'price_change_percentage_1y']])

df_market_data_scaled[:5]

array([[ 0.50852937,  0.49319307,  0.77220043,  0.23545963, -0.0674951 ,
        -0.35595348, -0.25163688],
       [ 0.18544589,  0.93444504,  0.55869212, -0.05434093, -0.27348273,
        -0.11575947, -0.19935211],
       [ 0.02177396, -0.70633685, -0.02168042, -0.06103015,  0.00800452,
        -0.55024692, -0.28206051],
       [-0.04076438, -0.81092807,  0.24945797, -0.05038797, -0.37316402,
        -0.45825882, -0.29554614],
       [ 1.19303608,  2.00095907,  1.76061001,  0.54584206, -0.29120287,
        -0.49984776, -0.27031695]])

In [130]:
# Create a DataFrame with the scaled data
df_market_data_norm = pd.DataFrame(df_market_data_scaled,columns =['price_change_percentage_24h', 'price_change_percentage_7d',\
       'price_change_percentage_14d', 'price_change_percentage_30d',\
       'price_change_percentage_60d', 'price_change_percentage_200d',\
       'price_change_percentage_1y'] )

# Copy the crypto names from the original data
df_market_data_norm["coin_id"] = df_market_data.index
# Set the coinid column as index
df_market_data_norm=df_market_data_norm.set_index("coin_id")

# Display sample data
df_market_data_norm.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [131]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))

In [132]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    model = KMeans(n_clusters=i, random_state=42)
    model.fit(df_market_data_norm)
    inertia.append(model.inertia_)


/Users/snehathomas/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/snehathomas/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/snehathomas/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/snehathomas/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.wa

In [133]:
# Create a dictionary with the data to plot the Elbow curve
elbow={"k":k,"inertia":inertia}

# Create a DataFrame with the data to plot the Elbow curve
elbow_df=pd.DataFrame(elbow)
elbow_df

k     inertia
0   1  287.000000
1   2  195.820218
2   3  123.190482
3   4   79.022435
4   5   65.220364
5   6   52.829819
6   7   43.786435
7   8   37.527740
8   9   33.070926
9  10   28.989073

In [134]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_df.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** At 4 clusters the inertia value does not drop much further and hits saturation. 

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [135]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4,random_state=42)

In [136]:
# Fit the K-Means model using the scaled data
model.fit(df_market_data_norm)

/Users/snehathomas/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


KMeans(n_clusters=4, random_state=42)

In [137]:
# Create a copy of the DataFrame
# Predict the clusters to group the cryptocurrencies using the scaled data
df_market_data_clusters = df_market_data_norm.copy()
# Add a new column to the DataFrame with the predicted clusters
df_market_data_clusters["cluster_prediction"]=model.predict(df_market_data_norm)

# Print the resulting array of cluster values.
# Display sample data
df_market_data_clusters

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

In [138]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
df_market_data_clusters.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="cluster_prediction",
    hover_cols = ["coin_id"], 
    title = "Scatter Plot by Coin Segment"
)

:NdOverlay   [cluster_prediction]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [139]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [140]:
# Use the PCA model with `fit_transform` to reduce to three principal components.
market_data_pca = pca.fit_transform(df_market_data_norm)
# View the first five rows of the DataFrame. 
market_data_pca[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [141]:
# Retrieve the explained variance to determine how much information can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The 3 pricipal components explains 89.5% of all variables - in this case it is split 37.2% , 34.7% and 17.6% across the 3 PCs

In [142]:
# Create a new DataFrame with the PCA data.
df_market_data_pca = pd.DataFrame(market_data_pca,columns=["PC1","PC2","PC3"])
# Creating a DataFrame with the PCA data

# Copy the crypto names from the original data
df_market_data_pca["coin_id"] = df_market_data.index

# Set the coinid column as index
df_market_data_pca = df_market_data_pca.set_index("coin_id")

# Display sample data
df_market_data_pca.head()

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [143]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1,11))

In [144]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    model = KMeans(n_clusters=i, random_state=42)
    model.fit(df_market_data_pca)
    inertia.append(model.inertia_)

/Users/snehathomas/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/snehathomas/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/snehathomas/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
/Users/snehathomas/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.wa

In [145]:
# Create a dictionary with the data to plot the Elbow curve
elbow_p={"k":k,"inertia_p":inertia}

# Create a DataFrame with the data to plot the Elbow curve
elbow_df_p=pd.DataFrame(elbow_p)
elbow_df_p

k   inertia_p
0   1  256.874086
1   2  168.811896
2   3   93.774626
3   4   49.665497
4   5   37.878747
5   6   27.618972
6   7   21.182776
7   8   17.091637
8   9   13.667065
9  10   10.559358

In [146]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_df_p.hvplot.line(
    x="k", 
    y="inertia_p", 
    title="Elbow Curve", 
    xticks=k
)

:Curve   [k]   (inertia_p)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best value for k is still 4 after PCA


  
* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No the vaue of k stays the same

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [147]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4,random_state=42)

In [148]:
# Fit the K-Means model using the PCA data
model.fit(df_market_data_pca)

/Users/snehathomas/anaconda3/lib/python3.10/site-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


KMeans(n_clusters=4, random_state=42)

In [149]:
# Create a copy of the DataFrame with the PCA data
# Predict the clusters to group the cryptocurrencies using the PCA data
df_market_data_pca_clusters = df_market_data_pca.copy()
# Add a new column to the DataFrame with the predicted clusters
df_market_data_pca_clusters["cluster_prediction_p"]=model.predict(df_market_data_pca)

# Print the resulting array of cluster values.
# Display sample data
df_market_data_pca_clusters

PC1       PC2       PC3  cluster_prediction_p
coin_id                                                                 
bitcoin              -0.600667  0.842760  0.461595                     3
ethereum             -0.458261  0.458466  0.952877                     3
tether               -0.433070 -0.168126 -0.641752                     0
ripple               -0.471835 -0.222660 -0.479053                     0
bitcoin-cash         -1.157800  2.041209  1.859715                     3
binancecoin          -0.516534  1.388377  0.804071                     3
chainlink            -0.450711  0.517699  2.846143                     3
cardano              -0.345600  0.729439  1.478013                     3
litecoin             -0.649468  0.432165  0.600303                     3
bitcoin-cash-sv      -0.759014 -0.201200 -0.217653                     0
crypto-com-chain     -0.248198 -1.376252 -1.462026                     0
usd-coin             -0.438408 -0.175337 -0.663388                     0
eos                  -0.693425 -0.473815 -0.527597                     0
monero                0.060499  2.909404  1.498571                     3
tron                 -0.393352 -0.108192 -0.012756                     0
tezos                -0.796176 -0.494409  1.082812                     3
okb                   0.064075 -1.269825 -1.098829                     0
stellar              -0.489015 -0.732719 -0.062543                     0
cosmos               -0.306272  0.703415  1.714224                     3
cdai                 -0.513528 -0.142802 -0.656566                     0
neo                  -0.362120 -0.986914 -0.728752                     0
wrapped-bitcoin      -0.604265  0.827398  0.439316                     3
leo-token            -0.413296 -0.674115 -1.076628                     0
huobi-token          -0.407483 -0.212507 -0.351426                     0
nem                   0.608974  0.563532 -1.148742                     0
binance-usd          -0.450211 -0.151019 -0.647401                     0
iota                 -0.764665 -0.517886  0.204990                     0
vechain              -0.556315 -1.938209 -1.261776                     0
zcash                -0.425147  0.492976  1.058048                     3
theta-token           2.676868 -0.013954 -1.965207                     0
dash                 -0.613923 -0.479337  0.339565                     0
ethereum-classic     -0.579924 -0.356334 -0.114942                     0
ethlend               8.089018 -3.896891  2.301382                     2
maker                -0.389045  0.165041  0.379414                     3
havven                0.865762 -2.261882  0.275583                     0
omisego               0.111675  0.428316 -1.205398                     0
celsius-degree-token  4.792395  6.767679 -1.986985                     1
ontology             -0.632355 -2.108117 -0.652227                     0
ftx-token            -0.593142  0.021485  0.209911                     0
true-usd             -0.458131 -0.135734 -0.635284                     0
digibyte             -0.297910 -0.191126 -0.909602                     0

In [150]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
df_market_data_pca_clusters.hvplot.scatter(
    x="PC1",
    y="PC2",
    by="cluster_prediction_p",
    hover_cols = ["coin_id"], 
    title = "Scatter Plot by Coin Segment (PCA)"
)

:NdOverlay   [cluster_prediction_p]
   :Scatter   [PC1]   (PC2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [151]:
# Composite plot to contrast the Elbow curves
# Did not use the + method for the elbow curves since this was more intuitive to compare. Did however use it for the scatter plots that cannot have the same axes.
elbow_composite_df = pd.merge(elbow_df,elbow_df_p, how ="inner", on = "k")
elbow_composite_df.hvplot.line(
    x="k", 
    y=["inertia","inertia_p"], 
    title="Elbow Curve Comparison", 
    xticks=k
)

:NdOverlay   [Variable]
   :Curve   [k]   (value)

In [158]:
# Composite plot to contrast the clusters

Original_scatter_plot = df_market_data_clusters.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="cluster_prediction",
    hover_cols = ["coin_id"], 
    title = "Scatter Plot by Coin Segment"
)
PCA_scatter_plot = df_market_data_pca_clusters.hvplot.scatter(
    x="PC1",
    y="PC2",
    by="cluster_prediction_p",
    hover_cols = ["coin_id"], 
    title = "Scatter Plot by Coin Segment (PCA)"
)

(Original_scatter_plot + PCA_scatter_plot).cols(1)

:Layout
   .NdOverlay.I  :NdOverlay   [cluster_prediction]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [cluster_prediction_p]
      :Scatter   [PC1]   (PC2,coin_id)

In [154]:
# Create a dataframe to understand the weights assigned to each variable for each principal component
feature_weights_df = pd.DataFrame(pca.components_, 
                                  columns=df_market_data_norm.columns, 
                                  index=['PCA' + str(i+1) for i in range(len(pca.components_))])

# Display the DataFrame
feature_weights_df

price_change_percentage_24h  price_change_percentage_7d  \
PCA1                    -0.416728                   -0.102432   
PCA2                     0.358162                    0.227432   
PCA3                    -0.218795                    0.787670   

      price_change_percentage_14d  price_change_percentage_30d  \
PCA1                    -0.009162                     0.191523   
PCA2                     0.540415                     0.562182   
PCA3                     0.349534                    -0.181744   

      price_change_percentage_60d  price_change_percentage_200d  \
PCA1                     0.320365                      0.594468   
PCA2                     0.434040                      0.027735   
PCA3                    -0.361377                      0.039733   

      price_change_percentage_1y  
PCA1                    0.568379  
PCA2                   -0.150789  
PCA3                    0.210541

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** The PCA allows us to explain most of the crptocurrencies with just two variables and is hence more efficient. It clearly shows two CCs 'ethlend' and 'cesius-degree-token' that lie in seperate clusters since they are on opposite ends of the values for PC2 and PC3